core/src/main/java/io/substrait/relation/ConsistentPartitionWindow.java

```diff
@@ -0,0 +1,70 @@
+package io.substrait.relation;
+
+import io.substrait.expression.Expression;
+import io.substrait.expression.Expression.SortField;
+import io.substrait.expression.FunctionArg;
+import io.substrait.expression.FunctionOption;
+import io.substrait.expression.WindowBound;
+import io.substrait.extension.SimpleExtension;
+import io.substrait.type.Type;
+import io.substrait.type.TypeCreator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+import org.immutables.value.Value;
+
+@Value.Immutable
+@Value.Enclosing
+public abstract class ConsistentPartitionWindow extends SingleInputRel implements HasExtension {
+
+  public abstract List<WindowRelFunctionInvocation> getWindowFunctions();
+
+  public abstract List<Expression> getPartitionExpressions();
+
+  public abstract List<SortField> getSorts();
+
+  @Override
+  protected Type.Struct deriveRecordType() {
+    Type.Struct initial = getInput().getRecordType();
+    return TypeCreator.of(initial.nullable())
+        .struct(
+            Stream.concat(
+                initial.fields().stream(),
+                getPartitionExpressions().stream().map(Expression::getType)));
+  }
+
+  @Override
+  public <O, E extends Exception> O accept(RelVisitor<O, E> visitor) throws E {
+    return visitor.visit(this);
+  }
+
+  public static ImmutableConsistentPartitionWindow.Builder builder() {
+    return ImmutableConsistentPartitionWindow.builder();
+  }
+
+  @Value.Immutable
+  public abstract static class WindowRelFunctionInvocation {
```

Comment:
Not for this PR. It might make sense in the future to define an interface of common types that we can share between `WindowRelFunctionInvocation` and `WindowFunctionInvocation`.

Constructing a `WindowRelFunctionInvocation` from that interface would be straightforward (as it would have all the fields), and then construction a `WindowRelFunctionInvocation` would just required providing the partitions and sorts.

We might also be able to have `WindowFunctionInvocation` extend from `WindowRelFunctionInvocation`, but I vaguely recall trying something like that in the past and the Immutable library not liking it.

Reply:
Yeah I agree -- it's subtly different and can be annoying to have concise builders and everything.